modules/tensor_mechanics/src/materials/EigenstrainHelper.C

```diff
@@ -0,0 +1,163 @@
+  applyEigenstrain(strain);
+}
+
+void
+EigenstrainHelper::applyTotalEigenstrain(
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains,
+    MaterialProperty<RankTwoTensor> & strain)
+{
+  if (eigenstrains.size() == 0)
+    return;
+  sumEigenstrainTotal(eigenstrains);
+  applyEigenstrain(strain);
+}
+
+void
+EigenstrainHelper::applyEigenstrain(MaterialProperty<RankTwoTensor> & strain)
+{
+  prepareEigenstrain();
+
+  for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+  {
+    RankTwoTensor fillme;
+    if (_eh_second_order)
+    {
+      for (unsigned i = 0; i < 6; ++i)
+      {
+        _eh_vals[i] = _eh_x[i](0);
+        for (unsigned j = 1; j < _eh_ncols; ++j)
+          _eh_vals[i] += _eh_x[i](j) * _eh_q_point[qp](j - 1);
+      }
+      _eh_adjusted_eigenstrain.fillFromInputVector(_eh_vals);
+    }
+    strain[qp] -= _eh_adjusted_eigenstrain;
+  }
+}
+
+void
+EigenstrainHelper::sumEigenstrainTotal(
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains)
+{
+  _eh_eigsum.resize(_eh_qrule->n_points());
+  for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+  {
+    _eh_eigsum[qp].zero();
+    for (auto es : eigenstrains)
+      _eh_eigsum[qp] += (*es)[qp];
+  }
+}
+
+void
+EigenstrainHelper::sumEigenstrainIncremental(
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains,
+    const std::vector<const MaterialProperty<RankTwoTensor> *> & eigenstrains_old)
+{
+  _eh_eigsum.resize(_eh_qrule->n_points());
+  for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+  {
+    _eh_eigsum[qp].zero();
+    for (unsigned i = 0; i < eigenstrains.size(); ++i)
+    {
+      _eh_eigsum[qp] += (*eigenstrains[i])[qp];
+      _eh_eigsum[qp] -= (*eigenstrains_old[i])[qp];
+    }
+  }
+}
+
+void
+EigenstrainHelper::prepareEigenstrain()
+{
+  // The eigenstrains can either be constant in an element or linear in x, y, z
+  // If constant, do volume averaging.
+  if (!_eh_second_order)
+  {
+    // Volume average
+    _eh_adjusted_eigenstrain.zero();
+    Real vol = 0.0;
+    for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+    {
+      _eh_adjusted_eigenstrain += _eh_eigsum[qp] * _eh_JxW[qp] * _eh_coord[qp];
+      vol += _eh_JxW[qp] * _eh_coord[qp];
+    }
+    _eh_adjusted_eigenstrain /= vol;
+  }
+  else
+  {
+    // 1 x y z
+
+    // Six sets (one for each unique component of eigen) of qp data
+    _eh_A.resize(_eh_qrule->n_points(), _eh_ncols);
+    for (auto && b : _eh_b)
+      b.resize(_eh_qrule->n_points());
+
+    for (unsigned qp = 0; qp < _eh_qrule->n_points(); ++qp)
+    {
+      _eh_A(qp, 0) = 1.0;
+      for (unsigned j = 1; j < _eh_ncols; ++j)
+        _eh_A(qp, j) = _eh_q_point[qp](j - 1);
+
+      _eh_b[0](qp) = _eh_eigsum[qp](0, 0);
+      _eh_b[1](qp) = _eh_eigsum[qp](1, 1);
+      _eh_b[2](qp) = _eh_eigsum[qp](2, 2);
+      _eh_b[3](qp) = _eh_eigsum[qp](1, 2);
+      _eh_b[4](qp) = _eh_eigsum[qp](0, 2);
+      _eh_b[5](qp) = _eh_eigsum[qp](0, 1);
```

Comment:
Hard coding to only six unique strains might set us up for problems later on if someone tries to code for shears that are not equal, although I understand wanting to avoid the extra computational effort.  Is there a way to efficiently check or guarantee for same/changing shears?

Reply:
If an approach is taken where stress and strain are not symmetric, I would imagine that you would need first to evaluate whether the issue we are tackling here is still an issue and then come up with a way to deal with it.  If the approach here makes sense, it could be generalized.  And yes, I am concerned that this will make the code noticeably slower.  We're not exactly speedy as it is.